Subject: Fix conflicting ncurses.h-COLOR_RED-define and gtest-COLOR_RED-enum

diff --git a/include/screen.hpp b/include/screen.hpp
index 377f7dd..2d4a943 100644
--- a/include/screen.hpp
+++ b/include/screen.hpp
@@ -37,4 +37,7 @@ int  screen_readkey();
 void screen_exit();
 void screen_show_errors();
 
+extern int key_backspace;
+extern int key_dc;
+
 #endif // __SCREEN_H__
diff --git a/include/sipp.hpp b/include/sipp.hpp
index 2eefe71..07442f5 100644
--- a/include/sipp.hpp
+++ b/include/sipp.hpp
@@ -55,9 +55,7 @@
 #include <map>
 #include <set>
 #include <math.h>
-#ifndef __SUNOS
-#include <curses.h>
-#else
+#ifdef __SUNOS
 #include <stdarg.h>
 #endif
 
diff --git a/src/screen.cpp b/src/screen.cpp
index 758f4bd..30c2372 100644
--- a/src/screen.cpp
+++ b/src/screen.cpp
@@ -40,6 +40,10 @@
 
 #include <unistd.h>
 
+/* Export these so others needn't include curses.h */
+int key_backspace = KEY_BACKSPACE;
+int key_dc = KEY_DC;
+
 unsigned long screen_errors;
 int           screen_inited = 0;
 char          screen_exename[255];
diff --git a/src/socket.cpp b/src/socket.cpp
index 0f4aae5..755a8ce 100644
--- a/src/socket.cpp
+++ b/src/socket.cpp
@@ -587,7 +587,7 @@ void handle_stdin_socket()
                 printf(SIPP_ENDL);
             }
 #ifndef __SUNOS
-            else if (c == KEY_BACKSPACE || c == KEY_DC)
+            else if (c == key_backspace || c == key_dc)
 #else
             else if (c == 14)
 #endif
